File: python/ml/verify.py
"""Verify provenance integrity of a saved model.

The provenance receipt is a lightweight chain of evidence:
split parameters → training fingerprint → assess ceremony status.

This is a receipt, not a blockchain. It catches accidental self-deception
(loading a model and re-assessing it, using test data from a different split)
rather than adversarial tampering.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any


def verify(path_or_model: str | Any) -> dict:
    """Verify provenance integrity of a model or saved model file.

    Checks:
    1. Provenance exists (model was trained through ml.split → ml.fit pipeline)
    2. Split receipt is present (cross-language verifiable split identity)
    3. Assess ceremony status (has it been assessed? how many times?)
    4. Chain integrity (training fingerprint → split lineage consistent)

    Parameters
    ----------
    path_or_model : str or Model
        Path to a .ml/.pyml file, or a fitted Model object.

    Returns
    -------
    dict
        Verification report with keys:
        - status: "verified" | "unverified" | "warning"
        - checks: list of individual check results
        - provenance: the raw provenance dict (if present)
        - assess_count: number of times assessed

    Examples
    --------
    >>> report = ml.verify("model.pyml")
    >>> report["status"]
    'verified'
    >>> report["assess_count"]
    1
    >>> ml.verify(model)["status"]
    'verified'
    """
    from ._types import Model, TuningResult

    checks = []

    # Load model if path
    if isinstance(path_or_model, (str, Path)):
        from .io import load
        model = load(str(path_or_model))
    elif isinstance(path_or_model, TuningResult):
        model = path_or_model.best_model
    elif isinstance(path_or_model, Model):
        model = path_or_model
    else:
        return {
            "status": "unverified",
            "checks": [{"check": "type", "ok": False,
                         "detail": f"Expected Model or path, got {type(path_or_model).__name__}"}],
            "provenance": None,
            "assess_count": 0,
        }

    provenance = getattr(model, "_provenance", None) or {}
    assess_count = getattr(model, "_assess_count", 0)

    # Check 1: Provenance exists
    has_provenance = bool(provenance.get("train_fingerprint"))
    checks.append({
        "check": "provenance_exists",
        "ok": has_provenance,
        "detail": "Training fingerprint recorded" if has_provenance
                  else "No provenance — model was not trained through ml.split → ml.fit pipeline",
    })

    # Check 2: Split receipt
    has_receipt = bool(provenance.get("split_receipt"))
    checks.append({
        "check": "split_receipt",
        "ok": has_receipt,
        "detail": f"Split receipt: {provenance['split_receipt']}" if has_receipt
                  else "No split receipt — model may predate v1.0 provenance or was trained on unsplit data",
    })

    # Check 3: Split lineage
    has_lineage = bool(provenance.get("split_id"))
    checks.append({
        "check": "split_lineage",
        "ok": has_lineage,
        "detail": f"Split ID: {provenance['split_id']}" if has_lineage
                  else "No split lineage recorded",
    })

    # Check 4: Assess ceremony
    if assess_count == 0:
        assess_detail = "Not yet assessed — test data untouched"
        assess_ok = True
    elif assess_count == 1:
        assess_detail = "Assessed once — ceremony completed correctly"
        assess_ok = True
    else:
        assess_detail = (
            f"Assessed {assess_count} times — one-shot ceremony violated. "
            "Results after the first assessment are statistically invalid."
        )
        assess_ok = False
    checks.append({
        "check": "assess_ceremony",
        "ok": assess_ok,
        "detail": assess_detail,
    })

    # Overall status
    all_ok = all(c["ok"] for c in checks)
    any_fail = any(not c["ok"] for c in checks)
    if all_ok:
        status = "verified"
    elif any_fail and has_provenance:
        status = "warning"
    else:
        status = "unverified"

    return {
        "status": status,
        "checks": checks,
        "provenance": provenance if has_provenance else None,
        "assess_count": assess_count,
    }

File: r/R/verify.R
#' Verify provenance integrity of a model
#'
#' Checks provenance chain: split parameters -> training fingerprint ->
#' assess ceremony status. Catches accidental self-deception (load-assess
#' loops, test-set shopping) rather than adversarial tampering.
#'
#' @param model An `ml_model`, `ml_tuning_result`, or path to `.mlr` file
#' @returns A list with `status` ("verified"/"unverified"/"warning"),
#'   `checks`, `provenance`, and `assess_count`.
#' @export
#' @examples
#' s <- ml_split(iris, "Species", seed = 42)
#' model <- ml_fit(s$train, "Species", seed = 42)
#' report <- ml_verify(model)
#' report$status
ml_verify <- function(model) {
  if (is.character(model)) {
    model <- ml_load(model)
  }
  if (inherits(model, "ml_tuning_result")) {
    model <- model[["best_model"]]
  }
  if (!inherits(model, "ml_model")) {
    return(list(
      status = "unverified",
      checks = list(list(check = "type", ok = FALSE,
                         detail = paste0("Expected ml_model, got ", class(model)[[1]]))),
      provenance = NULL,
      assess_count = 0L
    ))
  }

  provenance <- model[[".provenance"]]
  assess_count <- .get_assess_count(model)
  checks <- list()

  # Check 1: Provenance exists
  has_prov <- !is.null(provenance) && !is.null(provenance[["train_fingerprint"]])
  checks[[length(checks) + 1L]] <- list(
    check = "provenance_exists",
    ok = has_prov,
    detail = if (has_prov) "Training fingerprint recorded"
             else "No provenance -- model was not trained through ml_split -> ml_fit pipeline"
  )

  # Check 2: Split receipt
  has_receipt <- !is.null(provenance[["split_receipt"]])
  checks[[length(checks) + 1L]] <- list(
    check = "split_receipt",
    ok = has_receipt,
    detail = if (has_receipt) paste0("Split receipt: ", provenance[["split_receipt"]])
             else "No split receipt"
  )

  # Check 3: Split lineage
  has_lineage <- !is.null(provenance[["split_id"]])
  checks[[length(checks) + 1L]] <- list(
    check = "split_lineage",
    ok = has_lineage,
    detail = if (has_lineage) paste0("Split ID: ", provenance[["split_id"]])
             else "No split lineage recorded"
  )

  # Check 4: Assess ceremony
  if (assess_count == 0L) {
    assess_ok <- TRUE
    assess_detail <- "Not yet assessed -- test data untouched"
  } else if (assess_count == 1L) {
    assess_ok <- TRUE
    assess_detail <- "Assessed once -- ceremony completed correctly"
  } else {
    assess_ok <- FALSE
    assess_detail <- paste0("Assessed ", assess_count, " times -- one-shot ceremony violated")
  }
  checks[[length(checks) + 1L]] <- list(
    check = "assess_ceremony",
    ok = assess_ok,
    detail = assess_detail
  )

  # Overall status
  all_ok <- all(vapply(checks, function(c) c$ok, logical(1)))
  any_fail <- any(!vapply(checks, function(c) c$ok, logical(1)))
  status <- if (all_ok) "verified"
            else if (any_fail && has_prov) "warning"
            else "unverified"

  list(
    status = status,
    checks = checks,
    provenance = if (has_prov) provenance else NULL,
    assess_count = assess_count
  )
}
